src/main/scala/lms/core/codegen/cpp_codegen.scala

```diff
@@ -0,0 +1,102 @@
+package lms.core
+
+import lms.macros.RefinedManifest
+
+import scala.collection.mutable
+
+import java.io.{ByteArrayOutputStream, PrintStream}
+
+import Backend._
+import lms.core.stub._
+
+trait CPPOps { b: Base =>
+
+  abstract class Capture
+  object Capture {
+    def apply(s: String) = Seq(Share(s))
+    def CopyExcept[T:Manifest](s: Rep[T]*) = Seq(Share("=")) ++ s.map(UniqueRef(_))
+    def RefExcept[T:Manifest](s: Rep[T]*) = Seq(Share("&")) ++ s.map(UniqueCopy(_))
+  }
+  case class Share(s: String) extends Capture
+  case class UniqueCopy[T:Manifest](s: Rep[T]) extends Capture
+  case class UniqueRef[T:Manifest](s: Rep[T]) extends Capture
+  implicit def liftCapture(s:String) = Capture(s)
+  private def unwrap(c: Capture) = c match {
+    case Share(s: String) => lms.core.Backend.Const(s)
+    case UniqueCopy(s: Rep[_]) => Unwrap(s)
+    case UniqueRef(s: Rep[_]) => Unwrap(s)
+  }
```

Comment:
allowing programmer to explicitly control `capture` behavior (via copy or via reference)